test/email/sender_test.go

```diff
@@ -70,3 +74,78 @@ func TestSender(t *testing.T) {
 		t.Fatalf("Email not sent. Email: %+v\n", emails[0])
 	}
 }
+
+// TestContendingSenders ensures that each email generated by a cluster of
+// servers is sent exactly once. The test has several "servers" continuously
+// creating and "sending" emails.
+func TestContendingSenders(t *testing.T) {
+	ctx := context.Background()
+	logger := logrus.New()
+	db, err := database.New(ctx, test.DBTestCredentials(), logger)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if _, err = db.PurgeEmailCollection(ctx); err != nil {
+		t.Fatal("Failed to purge email collection:", err)
+	}
+	defer func() {
+		if _, err = db.PurgeEmailCollection(ctx); err != nil {
+			t.Fatal("Failed to purge email collection:", err)
+		}
+	}()
+	targetAddr := t.Name() + "@siasky.net"
+	numMsgs := 200
+	// count will hold the total number of messages sent.
+	var count int32
+	var wg sync.WaitGroup
+	// The generator will run in a thread and it will generate a predetermined
+	// number of messages.
+	generator := func(n int) {
+		defer wg.Done()
+		m := email.NewMailer(db)
+		for i := 0; i < n; i++ {
+			err1 := m.SendAddressConfirmationEmail(ctx, targetAddr, targetAddr)
+			if err1 != nil {
+				t.Fatal("Failed to send email.", err1)
```

Comment:
`Fatal` is illegal in goroutines. Usually `vet` complains about it cause it makes things difficult to debug sometimes.

So you usually want to use `t.Error(...)` followed by `return`.
Then later after `wg. Wait` you can check with `t.Failed()` if the test already failed and simply return if it did.

Reply:
Oh, right, I forgot about that. Thanks!